expression/builtin_control.go

```diff
@@ -127,34 +150,80 @@ func InferType4ControlFuncs(ctx sessionctx.Context, funcName string, lexp, rexp
 			resultFieldType.SetCollate(mysql.DefaultCollationName)
 			resultFieldType.SetFlag(0)
 		}
-		if evalType == types.ETDecimal || evalType == types.ETInt {
-			lhsUnsignedFlag, rhsUnsignedFlag := mysql.HasUnsignedFlag(lhs.GetFlag()), mysql.HasUnsignedFlag(rhs.GetFlag())
-			lhsFlagLen, rhsFlagLen := 0, 0
-			if !lhsUnsignedFlag {
-				lhsFlagLen = 1
-			}
-			if !rhsUnsignedFlag {
-				rhsFlagLen = 1
-			}
-			lhsFlen := lhs.GetFlen() - lhsFlagLen
-			rhsFlen := rhs.GetFlen() - rhsFlagLen
-			if lhs.GetDecimal() != types.UnspecifiedLength {
-				lhsFlen -= lhs.GetDecimal()
-			}
-			if lhs.GetDecimal() != types.UnspecifiedLength {
-				rhsFlen -= rhs.GetDecimal()
-			}
-			flen := maxlen(lhsFlen, rhsFlen) + resultFieldType.GetDecimal() + 1 // account for -1 len fields
-			resultFieldType.SetFlenUnderLimit(flen)
-		} else if evalType == types.ETString {
-			lhsLen, rhsLen := lhs.GetFlen(), rhs.GetFlen()
-			if lhsLen != types.UnspecifiedLength && rhsLen != types.UnspecifiedLength {
-				resultFieldType.SetFlen(mathutil.Max(lhsLen, rhsLen))
+	case ast.Case:
+		if len(args) == 0 {
+			panic("unexpected length 0 of args for casewhen")
+		}
+		fallthrough
+	default:
+		ec, err := CheckAndDeriveCollationFromExprs(ctx, funcName, evalType, args...)
+		if err != nil {
+			return err
+		}
+		resultFieldType.SetCollate(ec.Collation)
+		resultFieldType.SetCharset(ec.Charset)
+		resultFieldType.SetFlag(0)
+		for i := range args {
+			if !types.IsNonBinaryStr(args[i].GetType()) {
```

Comment:
according to the law of two types, the condition of add binary flag should be `if (!types.IsNonBinaryStr(args[i].GetType()) || mysql.HasBinaryFlag(args[i].GetType().GetFlag()))`?

Reply:
ok, updated.